### PR TITLE
core: add various stream utilities

### PR DESCRIPTION
**Pull request by frabert**
Useful for returning generators from collections and manipulating them in a lazy way, Haskell/LINQ style

### CODE REVIEW

File: core/include/gap/core/functional.hpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            for (auto& elem : enumerate(source)) {
+                if (pred(elem)) {
+                    co_yield elem;
+                }
+            }
+        }
+
+        template< typename TFunc, typename TSource >
+        auto flat_map(TFunc f, const TSource& source)
+            -> decltype(enumerate(f(*enumerate(source).begin()))) {
```
Comment (frabert) [inline core/include/gap/core/functional.hpp:54]: My metaprogramming-fu is not strong enough for me to figure out how to get rid of the ugly `decltype`s

Comment (lkorenc) [reply core/include/gap/core/functional.hpp:54]: Why just not leave it as `auto flat_map( ... ) { ` then?

File: core/include/gap/core/functional.hpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        template< typename TFunc, typename TSource >
+        auto map(TFunc f, const TSource& source)
+            -> gap::generator< decltype(f(*enumerate(source).begin())) > {
+            for (auto& elem : enumerate(source)) {
+                co_yield f(elem);
+            }
+        }
+
+        template< typename TPred, typename TSource >
+        auto filter(TPred pred, const TSource& source) -> decltype(enumerate(source)) {
```
Comment (lkorenc) [inline core/include/gap/core/functional.hpp:44]: Also I am not 100% sure why this (and others) does not have (also, not strictly instead) prototype that would take `filter( pred &&, auto &&range )` and then be called as `filter( pred, enumerate( x ) )`?

File: core/include/gap/core/functional.hpp (hunk 1 above)
Comment (frabert) [reply core/include/gap/core/functional.hpp:54]: Return type cannot be automatically inferred when using coroutines

File: core/include/gap/core/functional.hpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    #include "gap/core/generator.hpp"
+
+    #include <optional>
+
+namespace gap
+{
+    namespace functional
+    {
+        template< typename TContainer >
+        auto enumerate(const TContainer& cont) -> gap::generator< decltype(*cont.begin()) > {
```
[the remark this answers is not in the document]
Comment (frabert) [reply core/include/gap/core/functional.hpp]: Renamed to `stream`, taking a hint from Java

File: core/include/gap/core/functional.hpp (hunk 2 above)
Comment (frabert) [reply core/include/gap/core/functional.hpp]: I thought being able to `for(... : filter(pred, vector))` looked nicer than `for(... : filter(pred, enumerate(vector))`

Comment (lkorenc) [reply core/include/gap/core/functional.hpp]: Well you can technically have both.

Comment (frabert) [reply core/include/gap/core/functional.hpp]: opted to remove the implicit `stream` as it made things more difficult than necessary

File: core/include/gap/core/stream.hpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            -> gap::generator< decltype(f(*source.begin())) > {
+            for (auto&& elem : source) {
+                co_yield f(std::forward< decltype(elem) >(elem));
+            }
+        }
+
+        template< typename TPred, typename TElem >
+        gap::generator< TElem > filter(TPred&& pred, gap::generator< TElem > source) {
+            for (auto&& elem : source) {
+                if (pred(std::forward< decltype(elem) >(elem))) {
```
Comment (xlauko) [inline core/include/gap/core/stream.hpp:43]: forward here creates a dangling reference, you never want to move into predicate

**Review by xlauko — COMMENTED** [review]
Overall it feels like you are trying to adapt `std::views`.  In that case we want to be more aligned with `std` library.

**Comment by frabert** [comment]
Closing as `std::view` utilities would do the same things more elegantly

**Pull request closed without merging**